sqlglot/optimizer/qualify_columns.py

```diff
@@ -329,6 +338,47 @@ def _select_by_pos(scope: Scope, node: exp.Literal) -> exp.Alias:
         raise OptimizeError(f"Unknown output column: {node.name}")
 
 
+def _convert_columns_to_dots(scope: Scope, resolver: Resolver) -> None:
+    """
+    Converts `Column` instances that represent struct field lookup into chained `Dots`.
+
+    Struct field lookups look like columns (e.g. "struct"."field"), but they need to be
+    qualified separately and represented as Dot(Dot(...(<table>.<column>, field1), field2, ...)).
+    """
+    converted = False
+    for column in itertools.chain(scope.columns, scope.stars):
+        if isinstance(column, exp.Dot):
+            continue
+
+        column_table: t.Optional[str | exp.Identifier] = column.table
+        if (
+            column_table
+            and column_table not in scope.sources
```

Comment:
Could you please share your thought process here? As this is existing code we were reluctant in changing it along with the rest of the additions.

 Afaict `scope.selected_sources` is a subset of `scope.sources` and it's produced by "filtering"  the entries that are _not_ referenced in the scope, right? 

So in general, probing `scope.sources` might yield false positives for the scope?

Reply:
Thats exactly right, yeah

I'm OK if we don't want to change existing code. I didn't realize it was copy/pasted until halfway through my review.

Reply:
Not sure about the `selected_sources` here. I _think_ `scope.sources` was used because in lateral table explosions you won't see the preceding (exploded) tables in the UDTF's `scope.selected_sources`. They are, however, available in `scope.lateral_sources` (see [here](https://github.com/tobymao/sqlglot/blob/main/sqlglot/optimizer/scope.py#L688)), meaning that `scope.sources` also [contains](https://github.com/tobymao/sqlglot/blob/main/sqlglot/optimizer/scope.py#L73-L74) them.

With a quick check, there are a few tests that fail if we change this to `scope.selected_sources`, e.g.:

```diff
# execute: false
# dialect: redshift
- SELECT CAST(c.f AS VARCHAR(MAX)) AS f, e AS e FROM a.b AS c, c.d AS e
+ SELECT CAST(c.f AS VARCHAR(MAX)) AS f, e AS e FROM a.b AS c, c.c.d AS e
?                                                              ++
```